graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -718,6 +719,21 @@ def _compile_ast_node_to_ir(schema, current_schema_type, ast, location, context)
     return basic_blocks
 
 
+def _validate_all_tags_are_used(metadata):
+    """Ensure all tags are used in some filter."""
+    tag_names = set(metadata.tags)
+    filter_args = set()
+    for location, _ in metadata.registered_locations:
+        for filter_info in metadata.get_filter_infos(location):
+            for arg in filter_info.args:
+                if _is_tag_argument(arg):
+                    filter_args.add(arg[1:])  # Don't include % in tag name
+
+    unused_tags = tag_names - filter_args
+    if len(unused_tags) > 0:
+        raise GraphQLCompilationError(u'Found at least one unused @tag: {}.'.format(unused_tags))
```

Comment:
Good error messages are fundamental to a good user experience, and ones that are due to user error should always be actionable for the user -- they must know how to resolve their error. How about something like:
`This GraphQL query contains @tag directives whose values are not used: {}. This is not allowed. Please either use them in a filter, or remove them entirely.`